reps/2023-03-20-air-storage-path.md

```diff
@@ -0,0 +1,248 @@
+- In downstream components (`TrialRunner`, `Experiment`, `Trial`), we introduce respective arguments: `storage_path` and `experiment_path`
+
+We retain full backwards compatibility with the current API.
+
+## Synchronization
+
+The synchronization to a common location already works today. However, we can improve this
+for better efficiency and to better define the responsibilities of different components.
+
+### Current implementation
+
+The current flow multiplexes between two scenarios: With cloud storage enabled or without it.
+
+When cloud storage is enabled:
+
+- The driver saves experiment data to `/local/path/experiment_name`
+- The driver saves driver-based trial data to `/local/path/experiment_name/trial_id`
+- The driver uses a `Syncer` to periodically sync this state to the remote storage (e.g. S3), e.g. `s3://bucket/location/experiment_name`
+- This _includes_ driver-based trial data, such as the config and results log
+- This _excludes_ trainable-based trial data, such as checkpoints.
+  Uploading checkpoints is the trainable's responsibility.
+- The trainable saves trial-based data such as checkpoints to their local `/local/path/experiment_name/trial_id`
+- The trainable uses a `Syncer` to upload these checkpoints _synchronously_ to the remote storage after saving, e.g. `s3://bucket/location/experiment_name/trial_id/checkpoint_001`
+
+When cloud storage is disabled:
+
+- The driver saves experiment data to `/local/path/experiment_name`
+- The driver saves driver-based trial data to `/local/path/experiment_name/trial_id`
+- The trainable saves trial-based data such as checkpoints to their local `/local/path/experiment_name/trial_id`
+- The driver uses a `SyncerCallback` to synchronize contents form the trial directories (i.e. the missing trial-based data).
+  This uses the Ray object store.
+- This also happens _synchronously_ every time a trial reports a checkpoint has been saved
+
+### Future implementation
+
+In the first step (when we just change the API), this flow can remain the same, as it fulfills
+all requirements.
+
+The implementation for the case when a remote storage location is defined will remain as is.
+
+However, the implementation for the local storage path is currently inefficient because 
+the driver synchronously blocks until trial-based data is synchronized using the `SyncerCallback`.
+
+Instead, we should move this synchronization into the trial - analogous to the remote storage case.
+
+Thus, when a local storage path is defined (cloud storage is disabled):
+
+- The driver saves experiment data to `/local/path/experiment_name`
+- The driver saves driver-based trial data to `/local/path/experiment_name/trial_id`
+- The trainable saves trial-based data such as checkpoints to their local `/local/path/experiment_name/trial_id`
+- **The trainable uses the Ray object store to transfer these checkpoints _synchronously_ to the head node.**
```

Comment:
Who receives the data here, is there a separate actor that receives the transfer?

Reply:
Yes! Added a comment